zio-cache/shared/src/main/scala/zio/cache/Cache.scala

```diff
@@ -186,59 +174,149 @@ object Cache {
                   case MapValue.Complete(key, exit, _, timeToLive) =>
                     trackAccess(key)
                     trackHit()
-                    val now = Instant.now()
-                    if (now.isAfter(timeToLive)) {
+                    if (hasExpired(timeToLive)) {
                       map.remove(k, value)
                       get(k)
                     } else {
                       ZIO.done(exit)
                     }
+                  case MapValue.Refreshing(
+                        promiseInProgress,
+                        MapValue.Complete(mapKey, currentResult, _, ttl)
+                      ) =>
+                    trackAccess(mapKey)
+                    trackHit()
+                    if (hasExpired(ttl)) {
+                      promiseInProgress.await
+                    } else {
+                      ZIO.done(currentResult)
+                    }
                 }
               }
+            }
 
+          override def refresh(k: Key): IO[Error, Unit] =
+            ZIO.effectSuspendTotal {
+              for {
+                ref    <- Ref.make(map.get(k))
```

Comment:
There is no need for a `Ref` here since everything we are doing is on a single thread.

Reply:
@adamgfraser Is it because the entire block of `ZIO.effectSuspendTotal` runs on a single thread? I looked at its impl and java docs don't mention that.

Reply:
There isn't anything about `effectSuspendTotal` that does that you're just not doing anything concurrent here.

Reply:
Uh, I ran into some issues during the initial implementation. That was why I create a new promise (the question you asked below: https://github.com/zio/zio-cache/pull/52#discussion_r719054611), as opposed to the `var promise`... I'll revert to `var`'s (just like the `get` method) and make sure things are fine.

Reply:
I think the problem there was we had the race condition because you were forking a fiber to do the computation of the new value versus just computing it synchronously.

Reply:
Yeah possibly. I'll spend some time today to make sure things will be kosher. Thanks.